### PR TITLE
feat: add FaceNetSegmenter

### PR DESCRIPTION
**Pull request by strawberrypie**
This PR adds FaceNetSegmenter to be used in Jina. This class detects faces on an image and returns them as an NumPy array.  
This is a continuation of work being done in #6373.

Note that this segmenter could support batching, but other segmenters do not do that. I believe that it needs to be done in a separate PR.

### CODE REVIEW

File: segmenters/image/FaceNetSegmenter/__init__.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        :param kwargs: Additional positional arguments.
+        :return: A list with dicts that contain cropped images.
+        """
+        if self.channel_axis != self._default_channel_axis:
+            data = np.moveaxis(data, self.channel_axis, self._default_channel_axis)
+
+        with torch.no_grad():
+            images = torch.from_numpy(data.astype('float32')).to(self.device)
+            # Create a batch of size 1
+            images = images.unsqueeze(0)
```
Comment (nan-wang) [inline segmenters/image/FaceNetSegmenter/__init__.py:83]: To my understanding, you are dealing a single image at this point.
```suggestion
            image = torch.from_numpy(data.astype('float32')).to(self.device)
            # Create a batch of size 1
            image = image.unsqueeze(0)
```

File: segmenters/image/FaceNetSegmenter/__init__.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                faces = faces[0]
+                probabilities = probabilities[0]
+
+            faces = [(face, probability)
+                     for face, probability in zip(faces, probabilities)
+                     if face is not None]
+
+            results = [
+                dict(offset=0,
+                     weight=1.,
```
Comment (nan-wang) [inline segmenters/image/FaceNetSegmenter/__init__.py:98]: `weight` is used to denote the weights of each chunk and I think it fits to the concept of probability pretty well
```suggestion
                     weight=probability,
```

File: segmenters/image/FaceNetSegmenter/__init__.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                     blob=face.cpu().numpy(),
+                     tags={'probability': probability})
+                for face, probability in faces
+            ]
```
Comment (nan-wang) [inline segmenters/image/FaceNetSegmenter/__init__.py:102]: We can avoid looping the image for twice.
```suggestion
            results = []
            
            for face, probability in zip(faces, probabilities):
                if face is None:
                    continue
                results.append(dict(
                    offset=0,
                    weight=1.,
                    blob=face.detach().numpy(),
                ))
```

File: segmenters/image/FaceNetSegmenter/tests/test_facenetsegmenter.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                                      'four_faces.jpg',
+                                      'five_faces.jpg'])
+def test_segment_face(segmenter, filename):
+    image = Image.open(SEGMENTER_DIR / 'imgs' / filename)
+
+    result = segmenter.segment(np.array(image))
+
+    assert isinstance(result, list)
+    assert len(result) == 1
+    assert result[0]['blob'].shape == (3, 160, 160)
```
Comment (nan-wang) [inline segmenters/image/FaceNetSegmenter/tests/test_facenetsegmenter.py:30]: This is duplicated with the test `test_segment_face` at the bottom.

File: segmenters/image/FaceNetSegmenter/tests/test_facenetsegmenter.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    return FaceNetSegmenter(keep_all=True, post_process=False)
+
+
+@pytest.mark.parametrize('args', [
+    ('three_faces.jpg', 3),
+    ('four_faces.jpg', 4),
+    ('five_faces.jpg', 5)
+])
+def test_segment_face(segmenter_multiface, args):
+    filename, n_faces = args
```
Comment (nan-wang) [inline segmenters/image/FaceNetSegmenter/tests/test_facenetsegmenter.py:55]:
```suggestion
@pytest.mark.parametrize('filename, n_faces', [
    ('three_faces.jpg', 3),
    ('four_faces.jpg', 4),
    ('five_faces.jpg', 5)
])
def test_segment_face(segmenter_multiface, filename, n_faces):
```

File: segmenters/image/FaceNetSegmenter/tests/test_facenetsegmenter.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    image = Image.open(SEGMENTER_DIR / 'imgs' / filename)
```
Comment (nan-wang) [inline segmenters/image/FaceNetSegmenter/tests/test_facenetsegmenter.py:56]:
```suggestion
    import os
    image = Image.open(os.path.join(SEGMENTER_DIR, 'imgs', filename))
```

**Review by nan-wang — CHANGES_REQUESTED** [review]
Well done! I love this work!

File: segmenters/image/FaceNetSegmenter/__init__.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+        self.face_detector = MTCNN(selection_method=self.selection_method,
+                                   image_size=self.image_size,
+                                   margin=self.margin,
+                                   device=self.device,
+                                   post_process=self.post_process,
+                                   min_face_size=self.min_face_size,
+                                   keep_all=self.keep_all).eval()
+
+    def segment(self, data: 'np.ndarray', *args, **kwargs) -> List[Dict]:
```
Comment (nan-wang) [inline segmenters/image/FaceNetSegmenter/__init__.py:68]: This is to ensure every time only a sinle image is processed
```suggestion
    @single
    def segment(self, data: 'np.ndarray', *args, **kwargs) -> List[Dict]:
```

File: segmenters/image/FaceNetSegmenter/__init__.py (hunk 2 above)
Comment (strawberrypie) [reply segmenters/image/FaceNetSegmenter/__init__.py:98]: Sure, I considered using probability as weight, but wasn't sure that it was the right decision. Are there any docs on this?

File: segmenters/image/FaceNetSegmenter/tests/test_facenetsegmenter.py (hunk 4 above)
Comment (strawberrypie) [reply segmenters/image/FaceNetSegmenter/tests/test_facenetsegmenter.py]: Extracted this behavior to a function.

**Comment by strawberrypie** [comment]
Thanks for the review. I updated tests so that they behave correctly with the introduction of `@single`.

**Review by nan-wang — DISMISSED** [review]
LGTM👍

**Comment by nan-wang** [comment]
Ooops~Bad commit message, please refer to:
- [Commit Message Guideline for the First Time Contributor](https://github.com/jina-ai/jina/issues/553)
- [Contributing Guideline](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md)

The following commit message does not meet the guideline
```
Update segmenters/image/FaceNetSegmenter/__init__.py
```

**Comment by strawberrypie** [comment]
I fixed commit messages (they were auto-generated by GitHub) and force-pushed to this branch, but commit-linter still has a problem, probably due to force-push.

**Pull request merged**